woocommerce/woocommerce
# Code Freeze CLI: Add release changelog command

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:



Creates a changelog command that prepares a release branch and trunk at code freeze time. This CLI will replace [the current action](https://github.com/psealock/woocommerce/blob/trunk/.github/workflows/release-changelog.yml).



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites
1. You must have a `GITHUB_TOKEN` environment variable available in your PATH. This needs to be set to a [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) with repo permissions enabled.
2. Fork woocommerce onto your own namespace.

#### Prepare your fork
1. We'll pretend this is 7.8 code freeze, so update `trunk` to the latest.
2. Create a new branch, `release/7.8`, and push to your fork.
3. In case you're repeating this test, remove any branches that shouldn't be there
```
git push <remote> --delete update/7.8-changelog && git push <remote> --delete delete/7.8-changelog  
```
#### Test the command
1. Run the command pointing to your fork. The command uses a full clone with history, so it will take a couple minutes.
```
pnpm run utils code-freeze changelog -o <owner> -v 7.8
```
2. Once the command has completed, you will see two pull requests. For reference here are the same pull requests from the 7.7 release.
* https://github.com/woocommerce/woocommerce/pull/37776
* https://github.com/woocommerce/woocommerce/pull/37775
3. Confirm the pull requests created are identical to the 7.7 ones:
* base
* format
* content
* changelog format
* etc.
